### ISSUE TITLE
Problem with  `cachedEventIds` in the `fetchEvents` function

### ISSUE THREAD

**Issue by ashleyhood** [saade/filament-fullcalendar#22]
I am getting events disappearing when I use the `fetchEvents()` method to load calendar events.

If I comment out the line the caches the event ids in the `fullcalendar.blade.php` file, it works as expected.

```js
const fetchEvents = function ({ start, end }, successCallback, failureCallback) {
    @if( $this::canFetchEvents() )
        return $wire.fetchEvents({ start, end }, cachedEventIds)
            .then(events => {
                // Cache fetched events
                //cachedEventIds.push(...events.map(event => event.id)); <--- comment out this line fixes the issue

                return successCallback(events);
            })
            .catch( failureCallback );
    @else
        return successCallback([]);
    @endif
}
```

I think this is due to the way Fullcalendar calls Event Sources that are a function.

From the Fullcalendar docs:

> events (as a function) 
> ...
> FullCalendar will call this function whenever it needs new event data. This is triggered when the user clicks prev/next or switches views.

Is this line needed? I am not getting any duplication of events if it is commented out.

**Comment by saade** [comment]
hmm, can i see your implementation, please? We need to cache event ids that are returned from the fetchEvents to ignore them in the next fetchEvents call.

**Comment by ashleyhood** [comment]
Here is my implementation:

```php
public static function getAppointments(Carbon $start, Carbon $end, array $ignorableIds = []): array
{
    $appointments = Appointment::query()
        ->where([
            ['start_at', '>=', $start],
            ['end_at', '<=', $end],
        ])
        ->whereNotIn('id', $ignorableIds)
        ->get();

    return $appointments->map(function ($appointment) {
            return [
                'id' => $appointment->id,
                'title' => $appointment->title,
                'start' => $appointment->start_at,
                'end' => $appointment->end_at,
                'user_id' => $appointment->user_id,
                'client_id' => $appointment->client_id,
                'parent_id' => $appointment->parent_id,
                'location' => $appointment->location,
                'follow_up' => $appointment->follow_up,
                'type' => $appointment->type,
                'status' => $appointment->status,
            ];
        })
        ->toArray();
}

public function getViewData(): array
{
    return static::getAppointments(now()->subMonths(1), now()->addMonths(1));
}

public function fetchEvents(array $fetchInfo, array $ignorableIds): array
{
    $timezone = $this->config('timeZone', config('app.timezone'));

    $start = Carbon::parse($fetchInfo['start'], $timezone);
    $end = Carbon::parse($fetchInfo['end'], $timezone);

    return static::getAppointments($start, $end, $ignorableIds);
}
```

When I have the ids cached (`cachedEventIds.push(...events.map(event => event.id))`) and click the next month chevron and go 2 months forward and then back 1 month and then forward again (i.e. July->August->September->August->September), it won't show that months (i.e. September) events anymore.

**Comment by wychoong** [comment]
this is confirmed.

the reason is how FullCalendar handles previously loaded data when using function as `eventSources`

the issue is whenever FC trigger retrieves, it clears the old data, and when combine with the `$ignorableIds` in fetch, we lose the current events won't be able to fetch/display the events again.

there are 2 approaches to solve this:

1) not using `$ignorableIds`
this is the easiest. just remove the `cachedEventIds` from blade and `$ignorableIds` from method param

2) cache the events data fix in #24 
previously I included events data caching but removed by @saade . i forgot to raise this as i forgot about the reasoning behind it. So the purpose of caching the events data is to avoid the short blink of empty calendar when changing to month that is previously loaded (as FC clears all events, including days that span across month view). this cache provides better UX

**Comment by wychoong** [comment]
I further tested and figured that `ignorableIds` won't bring much benefits and it causes updated event info (if the info can change rather frequent)

so i combine the 2 approaches I mentioned in #25 by removing `ignorableIds` and implemented event data cache. caching on client side shouldn't be a big concern, or perhaps can have a config to turn it off.

@ashleyhood can you try if it solve your issue
